File: apps/api/src/internal/clinical/cognitive_anomaly_detector.ts
import "server-only";

import type { CognitiveBiomarkers } from "./cognitive_biomarkers";

/**
 * INTERNAL CLINICAL QUARANTINE — not for public API surfaces.
 * @module src/internal/clinical/cognitive_anomaly_detector
 */

export interface CognitiveAnomalyDetectorOptions {
  driftVarianceThreshold?: number;
  lexicalVarianceThreshold?: number;
}

export class CognitiveAnomalyDetector {
  readonly driftVarianceThreshold: number;
  readonly lexicalVarianceThreshold: number;

  constructor(options: CognitiveAnomalyDetectorOptions = {}) {
    this.driftVarianceThreshold = options.driftVarianceThreshold ?? 0.15;
    this.lexicalVarianceThreshold = options.lexicalVarianceThreshold ?? 0.1;
  }

  determineOverloadState(input: {
    current: CognitiveBiomarkers;
    baseline: CognitiveBiomarkers;
  }): boolean {
    const driftVariance = input.current.cohesionDrift - input.baseline.cohesionDrift;
    const lexicalVariance =
      input.baseline.lexicalDiversity - input.current.lexicalDiversity;

    return (
      driftVariance >= this.driftVarianceThreshold ||
      lexicalVariance >= this.lexicalVarianceThreshold
    );
  }
}

File: apps/mobile/retired_sprawl/lib_features/curiosity_loop/domain/services/cognitive_anomaly_detector.dart
import 'package:archiveme_mobile/features/curiosity_loop/domain/models/cognitive_biomarkers.dart';

/// Flags personalized cognitive drops relative to a user's EWMA baseline.
class CognitiveAnomalyDetector {
  const CognitiveAnomalyDetector({
    this.driftVarianceThreshold = 0.15,
    this.lexicalVarianceThreshold = 0.10,
  });

  static const double defaultDriftVarianceThreshold = 0.15;
  static const double defaultLexicalVarianceThreshold = 0.10;

  final double driftVarianceThreshold;
  final double lexicalVarianceThreshold;

  /// Returns true when current biomarkers indicate allostatic overload
  /// and grounding intervention is warranted.
  bool determineOverloadState({
    required CognitiveBiomarkers current,
    required CognitiveBiomarkers baseline,
  }) {
    final driftVariance = current.cohesionDrift - baseline.cohesionDrift;
    final lexicalVariance =
        baseline.lexicalDiversity - current.lexicalDiversity;

    return driftVariance >= driftVarianceThreshold ||
        lexicalVariance >= lexicalVarianceThreshold;
  }
}